### PR TITLE
Release

### PR DESCRIPTION
**Pull request by RobdeMooij**
First release version, now including:
- Readme with installation and getting started instructions.
- Demo notebooks with example outputs and visualizations.
- Example bare bones training script.
- Improved documentation.

### CODE REVIEW

**Comment by RobdeMooij** [comment]
Running checks first on a released version, running our scripts again on a fresh install to make sure everything works. Changes in functionality aren't expected, so please review for instructions/documentation/etc.

**Comment by RobdeMooij** [comment]
Demo's and training scripts all seem to work

**Comment by RobdeMooij** [comment]
Should `train_demo.py` be updated to make it for quicker checks whether everything works? 
E.g. to see quicker results:
- Printing/tqdm for batch loop
- Reduced number of epochs
- Smaller dataset (would have to generate a smaller demo split in `qcardia-data` demo?)

Since those were the things I changed before testing to not waste time...

**Comment by cianmscannell** [comment]
I guess it isn't strictly required but the requirements needs torchview to run the visualization

**Comment by cianmscannell** [comment]
In `qcardia-data` we assume the config is already in the working directory - i.e that the working directory is `demo` (`explore_config_path = Path("explore-config.yaml")`). 

But in `qcardia-models` we assume the working directory is above `demo` in `train_demo.py` (`config_path = Path("demo/train-demo-config.yaml")`)

Is this intentional because you expect to use script and notebooks differently?

**Comment by RobdeMooij** [comment]
> I guess it isn't strictly required but the requirements needs torchview to run the visualization

Yes, but it also needs Graphviz, both the Python package and the actual program on your system. As mentioned (pretty hidden imo) in their [installation instructions](https://github.com/mert-kurttutan/torchview?tab=readme-ov-file#installation):
> "For python interface of graphiz to work, you need to have dot layout command working in your system"

For Windows, this means installing Graphviz from [their website](https://www.graphviz.org/download/), with the executable location added to your Path. I think it works similarly on Linux/MacOS. The torchview installation instructions mention to use other package managers to install it, but I don't really like having to install even more stuff. Since I don't really like the torchview instructions we would also have to add instructions to install/use these package managers... 

Since it wouldn't work anyways with just additional torchview Python package requirements, I didn't include them at all and put a short explanation in the notebook. That explanation wasn't added to the public package version yet at the time of your comment, but should be there now. Maybe we should add some more specific instruction there? Would take more work and testing (for both of us) to have good and complete instructions for Windows, Linux, and MacOS.

**Comment by RobdeMooij** [comment]
> In `qcardia-data` we assume the config is already in the working directory - i.e that the working directory is `demo` (`explore_config_path = Path("explore-config.yaml")`).
> 
> But in `qcardia-models` we assume the working directory is above `demo` in `train_demo.py` (`config_path = Path("demo/train-demo-config.yaml")`)
> 
> Is this intentional because you expect to use script and notebooks differently?

Yes this is intentional, it is indeed a distinction between Python notebooks and scripts, not between packages.

**Comment by cianmscannell** [comment]
https://github.com/q-cardIA/qcardia-models/blob/b1b417d805290dc4d7b7c29824b9abe910c71586/src/qcardia_models/losses/dice_ce_loss.py#L74 

The code crashes on mac at this line. The reason is .get_device() returns 0 if GPU and -1 if CPU and not actually the name of device. If 0 PyTorch assumes CUDA even if it's not available. .device instead returns the actual device name. So changing to the below fixes it for me:

``` self.class_weights = self.class_weights.to(outputs.device) ```

@RobdeMooij can you check this doesn't break anything for you?

**Comment by RobdeMooij** [comment]
> https://github.com/q-cardIA/qcardia-models/blob/b1b417d805290dc4d7b7c29824b9abe910c71586/src/qcardia_models/losses/dice_ce_loss.py#L74
> 
> The code crashes on mac at this line. The reason is .get_device() returns 0 if GPU and -1 if CPU and not actually the name of device. If 0 PyTorch assumes CUDA even if it's not available. .device instead returns the actual device name. So changing to the below fixes it for me:
> 
> `self.class_weights = self.class_weights.to(outputs.device)`
> 
> @RobdeMooij can you check this doesn't break anything for you?

Still works for me

**Comment by cianmscannell** [comment]
Nice, then is the only thing missing to shorten the runtime of train_demo (in terms of iterations/epochs)? then I will do a final review

**Comment by RobdeMooij** [comment]
Limiting the number of batches in a dataloader is messy, so I hard forced it with `break` and `enumerate`. Since `tqdm` doesn't work with this hard force, `tqdm` is used on the 10 epochs, all with very few batches. The training is now heavily bottlenecked by the dataloading, so the first epoch takes far longer than the others, but the training should still finish in <1 minute.

**Comment by cianmscannell** [comment]
also works

File: README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+pip install git+https://github.com/q-cardIA/qcardia-models@v1.0.0
+```
+
+Available versions can be found in the [releases tab](https://github.com/q-cardIA/qcardia-models/releases), where the release [tags](https://github.com/q-cardIA/qcardia-models/tags) are used in the install command.
+
+#### Editable version of `qcardia-models`
+You can install a local copy in `editable` mode to make changes to the package that instantly get reflected in your environment. After getting a local copy (download/clone/fork), install using:
+```
+pip install -e "path/to/qcardia-models"
+```
```
Comment (cianmscannell) [inline README.md:31]: You need to have a recent version of pip for this to work. Old versions required a `setup.py`.
So I like to add also the disclaimer:

(It may be neccessary to first upgrade pip using `python3 -m pip install --upgrade pip`).

File: README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    mlp_channels_list=[512, 512, 512],
+)
+```
+
+You can take a look in the demo folder for the included demo notebooks for more information and examples. Tutorials for each component of `qcardia-models` in isolation can be found in the `demo/tutorial.ipynb` notebook. The `demo/model_visualization.ipynb` notebook visualizes two available models and their components (requires the `torchview` package, see notebook for instructions).
+
+The `train_demo.py` training script showcases a bare-bones example of training a model. The training demo uses our [qcardia-data](https://github.com/q-cardIA/qcardia-data) package to get training data, but other data(loaders) can also be used. To find the config files, the training demo script assumes a main/root package folder level current working directory (i.e. parent folder of demo folder), while the demo notebooks assume the demo folder (where the notebooks are located) as the current working directory. Make sure to update your file/folder hierarchy and current working directories when running scripts/notebooks that cannot find config files, or change config paths in the code directly to reflect the correct (global) path of the config.
```
Comment (cianmscannell) [inline README.md]: Call it `demo/train_demo.py` for consistency with the previous paragraph

File: demo/encodermlp-config.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  - 512
+mlp:
+  channels_list:
+  - 512
+  - 512
+  - 512
+
+
+loss:
+  ntxent_temperature: 0.1
```
Comment (cianmscannell) [inline demo/encodermlp-config.yaml:19]: Is it weird that we have a general model repo that introduce an encoder-mlp but already assumes a contrastive loss? The tutorial part for this could just use normal cross entropy or even MSE?

File: demo/encodermlp-config.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  ntxent_supervised_threshold: 0.1
+  ntxent_supervised_cyclic_labels: False
+
+
+training:
+  nr_epochs_frozen_encoder: 0
```
Comment (cianmscannell) [inline demo/encodermlp-config.yaml]: also this, is it used in the tutorial/demo?

**Review by cianmscannell — APPROVED** [review]

File: demo/encodermlp-config.yaml (hunk 3 above)
Comment (RobdeMooij) [reply demo/encodermlp-config.yaml:19]: I don't think it's weird, since the contrastive loss is part of the models repo/package? The tutorial showcases how to use this loss function.

File: demo/encodermlp-config.yaml (hunk 4 above)
Comment (RobdeMooij) [reply demo/encodermlp-config.yaml]: I agree this can be removed, since it's never used for pretraining
